# stLFR to Supernova

## BRIEF
__________________

- The stLFR2Supernova is a pipeline which we can de-novo-assemble the stLFR reads by Supernova Assembler. 
    -  the stLFR reads means the raw reads generated by stLFR technology from BGI[1].
    -  the Supernova Assembler means the de-novo software from 10X Genomes [2].

## INSTALL
__________________

### how to download

```
git clone https://gitlab.com/BGIQD/stlfr2supernova_pipeline.git YOUR-INSTALL-DIR
```

### how to install

Becasuse it is writed by perl and shell script, there is no extra need to install or compeir this pipeline.

## USAGE
__________________

### structure of the files

```
├── README.md
├── bin
│   ├── SOAPfilter_v2.2         # used to filter dup & adaptor.
│   ├── split_barcode.pl        # split barcode sequence into barcode number.
│   ├── merge_barcodes.pl       # random merge 8 stLFR barcode number into 1 10X barcode.
│   └── fake_10x.pl             # convert format from stLFR reads to 10X reads.
├── data
│   ├── barcode.list            # the barcode list of stLFR.
│   └── lane.lst                # config for SOAPfilter.
├── profile                     # the default profile.
├── run.sh                      # one step pipeline,will call step 0 ~ 3
├── step_0_split_barode.sh      # call split_barcode.pl and generate split_reads.*.fq.gz & barcode_freq.txt
├── step_1_filter.sh            # call SOAPfilter_v2.2 and generate split_reads.*.fq.gz.clean.gz
├── step_2_fake_10X_data.sh     # call merge_barcodes.pl & fake_10x.pl and generate merge.txt & read-R*_si-TTCACGCG_lane-001-chunk-001.fastq.gz
└── step_3_run_supernova.sh     # call Supernova Assmebler
```


### step by step usage.

- 1th, create a new working folder

```
> mkdir YourProjectRoot
```

- 2th, copy the default profile into your working folder

```
> cd YourProjectRoot
> cp YOUR-INSTALL-DIR/profile ./your_own_profile
```

- 3th, edit your_own_profile

```
> vi ./your_own_profile 
```
- 4.0th, run the whole pipeline with your profile

```
> YOUR-INSTALL-DIR/run.sh  # make sure your run this command in YourProjectRoot
```
- 4.1th, if you don't want to run the whole pipeline, you can run it step by step and only run what you need .

```
> YOUR-INSTALL-DIR/step_x_xxxx.sh # make sure your run this command in YourProjectRoot
```

### how to configure the profile 

TODO

## OTHERS
__________________

### how to re-generate the barcode_freq.txt by split_reads.1.fq.gz

```
gzip -dc split_reads.1.fq.gz | grep '@CL' | awk -F '[# |]' '{print$2}' | awk -F '/' '{print $1}' | sort | uniq -c | awk '{printf("%s\t%s\n",$2,$1);}' > ~/barcode_freq.txt
```

## REFRENCE
__________________

[1] [Efficient and unique co-barcoding of second-generation sequencing reads from long DNA molecules enabling cost effective and accurate sequencing, haplotyping, and de novo assembly][11]
 
[2] [Direct determination of diploid genome sequences][22]

[11]: https://www.ncbi.nlm.nih.gov/pubmed/30940689 
[22]: https://www.ncbi.nlm.nih.gov/pubmed/28381613 
