# stLFR to Supernova

## <a name=intro></a> Introduction

- The stLFR2Supernova is a pipeline which we can de-novo-assemble the stLFR reads by Supernova Assembler. 
    -  the stLFR reads means the raw reads generated by stLFR technology from BGI [1].
    -  the Supernova Assembler means the de-novo software from 10X Genomes [2].

## <a name=contents></a> Table of Contents

- [Introduction](#intro)
- [Table of Contents](#contents)
- [User's Guide](#user-guide)
    - [Installation](#install)
    - [Structure of the files](#files)
    - [General usage](#usage)
    - [The profile file](#profile)
- [Use cases](#use-cases)
- [Miscellaneous](#misc)
- [Reference](#ref)

## <a name=user-guide></a> User's Guide

### <a name=install></a>Installation

```
git clone https://gitlab.com/BGIQD/stlfr2supernova_pipeline.git YOUR-INSTALL-DIR
```

* *It is wrote by perl and shell script.*
* *No need to do extra compiling or installing.*

### <a name=files></a> Structure of the files

```
├── README.md
├── bin
│   ├── SOAPfilter_v2.2         # filter dup & adaptor.
│   ├── split_barcode.pl        # split barcode sequence into barcode number.
│   ├── merge_barcodes.pl       # random map 8 stLFR barcode to 1 10X barcode.
│   └── fake_10x.pl             # convert stLFR reads to 10X format reads.
├── data
│   ├── barcode.list            # the barcode list of stLFR.
│   └── lane.lst                # config for SOAPfilter.
├── profile                     # the default profile.
├── run.sh                      # one main step pipeline,which call four substeps 0 ~ 3
├── step_0_split_barode.sh      # call split_barcode.pl and generate split_reads.*.fq.gz & barcode_freq.txt
├── step_1_filter.sh            # call SOAPfilter_v2.2 and generate split_reads.*.fq.gz.clean.gz
├── step_2_fake_10X_data.sh     # call merge_barcodes.pl & fake_10x.pl and generate merge.txt & read-R*_si-TTCACGCG_lane-001-chunk-001.fastq.gz
└── step_3_run_supernova.sh     # call Supernova Assmebler
```

### <a name=usage></a> General usage

- 1st, create a new working folder

```
> mkdir YourProjectRoot
```

- 2nd, copy the default profile into your working folder

```
> cd YourProjectRoot
> cp YOUR-INSTALL-DIR/profile ./your_own_profile
```
    **Do not modify the filename.Keep it as "profile"**

- 3rd, edit your_own_profile

```
> vi ./your_own_profile 
```
- 4th, run the main pipeline with your profile or run it step by step

    - 4.1 run the main pipeline by one step
```
> YOUR-INSTALL-DIR/run.sh  # make sure your run this command in YourProjectRoot
```
    - 4.2 run the pipeline step by step or only run what you need

```
> YOUR-INSTALL-DIR/step_x_xxxx.sh # make sure your run this command in YourProjectRoot
```

### <a name=profile></a> The profile file

**Do not modify the filename.Keep it as "profile".**

**Make sure the profile file is in YourProjectRoot directory.**

```
#
#   basic settings below 
#  MODIFY HERE FOR YOU PROJECT !!!
#
r1="L1.1.fq.gz L2.1.fq.gz"  # the stLFR raw read1. differnt lane use " " to seperate.
r2="L1.2.fq.gz L2.2.fq.gz"  # the stLFR raw read1. differnt lane use " " to seperate.
USE_FILTER="yes"            # yes or no. # use SOAPFilter or not
# below are baisc parameters for supernova assembler
PROJECT_NAME="Human"        # supernova's --id
THREADS=80                  # supernova's --localcores
MEMORY=100 #GB              # supernova's --localmem
MAX_READS=1200000000        # supernova's --maxreads #    for supernova version <= 2.10
#MAX_READS=2140000000 #for supernova 2.11

#
#   exec path below
#
#  MODIFY HERE FOR YOU ENVIROMENT !!!

# "SCRIPT_PATH" is the YOUR-INSTALL-DIR directory .
SCRIPT_PATH="~/software/stlfr2supernova/"
#"SUPERNOVA" is the executing directory of supernova. Different version number can be accepted.
SUPERNOVA="~/software/supernova-2.0.0/"  
#"SOAP_FILTER is the executing diretory of SOAPFilter
SOAP_FILTER="/hwfssz1/ST_OCEAN/USER/guolidong/stLFR/data_pipeline/SOAPfilter_v2.2.1/SOAPfilter_v2.2"

#
#   intermediate files that will be generated/needed below .
#       DO NOT MODIFY BELOW
# UNLESS YOU KNOW WHAT YOU ARE DOING.
#
R1="tmp_r1.fq.gz"                   # A symbol-link if raw reads only contain 1 lane, a concatenate of all lane otherwise.
R2="tmp_r2.fq.gz"
SPLIT="split_reads"                 # the prefix of splited reads. split_reads.1.fq.gz & split_reads.2.fq.gz
BARCODE_FREQ="./barcode_freq.txt"   # the barcode frequence information
MERGE="merge.txt"                   # the stLFR->10X barcode mapping information 
SUPERNOVA_R1="read-R1_si-TTCACGCG_lane-001-chunk-001.fastq.gz" # the output 10X reads
SUPERNOVA_I1="read-I1_si-TTCACGCG_lane-001-chunk-001.fastq.gz" # the output 10X reads
SUPERNOVA_R2="read-R2_si-TTCACGCG_lane-001-chunk-001.fastq.gz" # the output 10X reads

```

## <a name=use-cases></a> Use cases

- If there is no barcode_freq.text file in the directory "YourProjectRoot", we can re-generate the barcode_freq.txt by split_reads.1.fq.gz as following command 

```
gzip -dc split_reads.1.fq.gz | awk '!(NR%4-1)' | awk -F '[# |]' '{print$2}' | awk -F '/' '{print $1}' | sort | uniq -c | awk '{printf("%s\t%s\n",$2,$1);}' > barcode_freq.txt
```

## <a name=misc></a> Miscellaneous

- Requirements
    - Linux system && Bash
    - Perl
    - Supernova Assembler
    - SOAPFilter ( optional )
- Limitations
    - What Supernova Assmebler can't do, we also can't .
- Resources
    - There are two step that may cause huge memory cost:
        - step_0_split_barode.sh . this cost depends on your raw reads.
        - step_3_run_supernova.sh. this cost can be limited by MEMORY ( in profile ).
    - Only step_3_run_supernova.sh suport multi-threads ( by Supernova ).

## <a name=ref></a> Reference

[1] [Efficient and unique co-barcoding of second-generation sequencing reads from long DNA molecules enabling cost effective and accurate sequencing, haplotyping, and de novo assembly][11]
 
[2] [Direct determination of diploid genome sequences][22]

[11]: https://www.ncbi.nlm.nih.gov/pubmed/30940689 
[22]: https://www.ncbi.nlm.nih.gov/pubmed/28381613 

